src/status_emitter/json.rs

```
@@ -0,0 +1,157 @@
+#[derive(Clone)]
+pub struct JSON;
+
+impl StatusEmitter for JSON {
+    /// Create a report about the entire test run at the end.
+    fn finalize(
+        &self,
+        failed: usize,
+        succeeded: usize,
+        ignored: usize,
+        filtered: usize,
+        aborted: bool,
+    ) -> Box<dyn Summary> {
+        let status = if aborted || failed > 0 {
+            String::from("failed")
+        } else {
+            String::from("ok")
+        };
+
+        emit_suite_end(failed, filtered, ignored, succeeded, &status);
+
+        Box::new(())
+    }
+
+    /// Invoked the moment we know a test will later be run.
+    /// Useful for progress bars and such.
+    fn register_test(&self, path: PathBuf) -> Box<dyn TestStatus + 'static> {
+        let name = path.to_str().unwrap().to_string();
+
+        emit_test_start(&name);
+
+        Box::new(JSONStatus {
+            name,
+            path,
+            revision: String::new(),
+            style: RevisionStyle::Show,
+        })
+    }
+}
+
+/// Information about a specific test run.
+pub struct JSONStatus {
+    name: String,
+    path: PathBuf,
+    revision: String,
+    style: RevisionStyle,
+}
+
+impl TestStatus for JSONStatus {
+    /// A test has finished, handle the result immediately.
+    fn done(&self, result: &TestResult, aborted: bool) {
+        let status = if aborted {
+            String::from("timeout")
+        } else {
+            match result {
+                Ok(TestOk::Ignored) => String::from("ignored"),
+                Ok(TestOk::Ok) => String::from("ok"),
+                Err(_) => String::from("failed"),
+            }
+        };
+        let error_output = if let Err(errored) = result {
+            let command = &errored.command;
+            let stderr = errored.stderr.to_str_lossy();
+            let stdout = errored.stdout.to_str_lossy();
+            format!(r#"---- command\n\n{command}\n\n\---- stdout\n\n\{stdout}\n\n---- stderr\n\n\{stderr}"#)
```

huh, test failures are not reported in json format?

---

oh nevermind 🤦 that's put into json